Lab Assignment 2 - Sentiment Analysis

Author Name : maurya sasanka Bhima

ASU ID : 1234108592

File Creation Date : 02/02/2025

In [18]:
#  Import Necessary Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Load the restaurant review data
file_path = '/content/restaurant_reviews_az.csv'
data = pd.read_csv(file_path)

# Displaying the first few rows of the dataset
print("Sample Data from the Dataset:")
print(data.head())

# Showing basic dataset information
print("\nDataset Information:")
print(data.info())

# Displaying column names
print("\nColumn Names:", data.columns.tolist())

Sample Data from the Dataset:
                review_id                 user_id             business_id  \
0  IVS7do_HBzroiCiymNdxDg  fdFgZQQYQJeEAshH4lxSfQ  sGy67CpJctjeCWClWqonjA   
1  QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2  oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3  E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
4  Rd222CrrnXkXukR2iWj69g  LPxuausjvDN88uPr-Q4cQA  CA5BOxKRDPGJgdUQ8OUOpw   

   stars  useful  funny  cool  \
0      3       1      1     0   
1      5       1      1     1   
2      5       1      0     0   
3      5       0      0     0   
4      4       1      0     0   

                                                text                 date  
0  OK, the hype about having Hatch chili in your ...  2020-01-27 22:59:06  
1  Pandemic pit stop to have an ice cream.... onl...  2020-04-19 05:33:16  
2  I was lucky enough to go to the soft opening a...  2020-02-29 19:43:44  
3  I

In [21]:
# Remove 3-star reviews
filtered_data = data[data['stars'] != 3].copy()

# Create the Sentiment column: 0 for 1-2 stars, 1 for 4-5 stars
filtered_data['Sentiment'] = np.where(filtered_data['stars'] <= 2, 0, 1)

# Display the updated dataset
print("Sample Data After Filtering:")
print(filtered_data.head())

# Show the value counts for Sentiment column to verify the distribution
print("\nSentiment Distribution:")
print(filtered_data['Sentiment'].value_counts())

Sample Data After Filtering:
                review_id                 user_id             business_id  \
1  QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2  oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3  E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
4  Rd222CrrnXkXukR2iWj69g  LPxuausjvDN88uPr-Q4cQA  CA5BOxKRDPGJgdUQ8OUOpw   
5  kx6O_lyLzUnA7Xip5wh2NA  YsINprB2G1DM8qG1hbrPUg  rViAhfKLKmwbhTKROM9m0w   

   stars  useful  funny  cool  \
1      5       1      1     1   
2      5       1      0     0   
3      5       0      0     0   
4      4       1      0     0   
5      1       0      0     0   

                                                text                 date  \
1  Pandemic pit stop to have an ice cream.... onl...  2020-04-19 05:33:16   
2  I was lucky enough to go to the soft opening a...  2020-02-29 19:43:44   
3  I've gone to claim Jumpers all over the US and...  2020-03-14 21:47:07   
4

In [22]:
# Process the data and prepare training and test sets
# Assume the review text is stored in a column named 'text'
X = data['text']
y = data['Sentiment']

# Split the data into training (80%) and test (20%) sets, with a fixed random state for reproducibility
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

# Print the shapes of the resulting datasets
print("Training set size:", X_train.shape, y_train.shape)
print("Test set size:", X_test.shape, y_test.shape)


Training set size: (35274,) (35274,)
Test set size: (8819,) (8819,)


In [6]:
#  Represent documents using CountVectorizer (frequency count representation)
# Initialize CountVectorizer with a maximum of 1000 features
count_vect = CountVectorizer(max_features=1000)

# Fit CountVectorizer on the training data and transform both training and test data
X_train_count = count_vect.fit_transform(X_train)
X_test_count = count_vect.transform(X_test)

# Display the shapes of the transformed datasets
print("CountVectorizer - Training data shape:", X_train_count.shape)
print("CountVectorizer - Test data shape:", X_test_count.shape)


CountVectorizer - Training data shape: (35274, 1000)
CountVectorizer - Test data shape: (8819, 1000)


In [24]:
#  Train a Naive Bayes classifier using CountVectorizer features and evaluate its performance

# Initialize the Multinomial Naive Bayes classifier
nb_classifier_count = MultinomialNB()

# Train the classifier on the count vectorized training data
nb_classifier_count.fit(X_train_count, y_train)

# Predict sentiments for the test data
y_pred_nb_count = nb_classifier_count.predict(X_test_count)

# Evaluate the classifier's performance
accuracy_nb_count = accuracy_score(y_test, y_pred_nb_count)
print("Naive Bayes (CountVectorizer) Accuracy:", accuracy_nb_count)
print("Classification Report:\n", classification_report(y_test, y_pred_nb_count))


Naive Bayes (CountVectorizer) Accuracy: 0.9209660959292437
Classification Report:
               precision    recall  f1-score   support

           0       0.87      0.86      0.86      2554
           1       0.94      0.95      0.94      6265

    accuracy                           0.92      8819
   macro avg       0.91      0.90      0.90      8819
weighted avg       0.92      0.92      0.92      8819



In [8]:
#  Train an SVM classifier using CountVectorizer features and evaluate its performance

# Initialize a Linear SVM classifier with a fixed random state for reproducibility
svm_classifier_count = LinearSVC(random_state=42)

# Train the SVM classifier on the count vectorized training data
svm_classifier_count.fit(X_train_count, y_train)

# Predict sentiments for the test data using the SVM classifier
y_pred_svm_count = svm_classifier_count.predict(X_test_count)

# Evaluate the classifier's performance
accuracy_svm_count = accuracy_score(y_test, y_pred_svm_count)
print("SVM (CountVectorizer) Accuracy:", accuracy_svm_count)
print("Classification Report:\n", classification_report(y_test, y_pred_svm_count))


SVM (CountVectorizer) Accuracy: 0.9477264996031296
Classification Report:
               precision    recall  f1-score   support

           0       0.92      0.90      0.91      2554
           1       0.96      0.97      0.96      6265

    accuracy                           0.95      8819
   macro avg       0.94      0.93      0.94      8819
weighted avg       0.95      0.95      0.95      8819



In [9]:
#  Represent documents using the TF-IDF vectorizer
# Initialize TF-IDF Vectorizer with a maximum of 1000 features
tfidf_vect = TfidfVectorizer(max_features=1000)

# Fit and transform the training data using TF-IDF and transform the test data
X_train_tfidf = tfidf_vect.fit_transform(X_train)
X_test_tfidf = tfidf_vect.transform(X_test)

# Print the shapes of the TF-IDF transformed datasets
print("TF-IDF - Training data shape:", X_train_tfidf.shape)
print("TF-IDF - Test data shape:", X_test_tfidf.shape)


TF-IDF - Training data shape: (35274, 1000)
TF-IDF - Test data shape: (8819, 1000)


In [10]:
#  Train a Naive Bayes classifier using TF-IDF features and evaluate its performance

# Initialize the Multinomial Naive Bayes classifier
nb_classifier_tfidf = MultinomialNB()

# Train the classifier on the TF-IDF training data
nb_classifier_tfidf.fit(X_train_tfidf, y_train)

# Predict sentiments for the test data using the TF-IDF features
y_pred_nb_tfidf = nb_classifier_tfidf.predict(X_test_tfidf)

# Evaluate the classifier's performance
accuracy_nb_tfidf = accuracy_score(y_test, y_pred_nb_tfidf)
print("Naive Bayes (TF-IDF) Accuracy:", accuracy_nb_tfidf)
print("Classification Report:\n", classification_report(y_test, y_pred_nb_tfidf))


Naive Bayes (TF-IDF) Accuracy: 0.9070189363873455
Classification Report:
               precision    recall  f1-score   support

           0       0.94      0.73      0.82      2554
           1       0.90      0.98      0.94      6265

    accuracy                           0.91      8819
   macro avg       0.92      0.85      0.88      8819
weighted avg       0.91      0.91      0.90      8819



In [11]:
#  Train an SVM classifier using TF-IDF features and evaluate its performance

# Initialize a Linear SVM classifier for TF-IDF features
svm_classifier_tfidf = LinearSVC(random_state=42)

# Train the SVM classifier on the TF-IDF training data
svm_classifier_tfidf.fit(X_train_tfidf, y_train)

# Predict sentiments for the test data using the SVM classifier
y_pred_svm_tfidf = svm_classifier_tfidf.predict(X_test_tfidf)

# Evaluate the classifier's performance
accuracy_svm_tfidf = accuracy_score(y_test, y_pred_svm_tfidf)
print("SVM (TF-IDF) Accuracy:", accuracy_svm_tfidf)
print("Classification Report:\n", classification_report(y_test, y_pred_svm_tfidf))


SVM (TF-IDF) Accuracy: 0.9499943304229505
Classification Report:
               precision    recall  f1-score   support

           0       0.92      0.91      0.91      2554
           1       0.96      0.97      0.96      6265

    accuracy                           0.95      8819
   macro avg       0.94      0.94      0.94      8819
weighted avg       0.95      0.95      0.95      8819



In [13]:
#  Use the VADER sentiment analyzer to predict review sentiment and evaluate its performance

# Initialize the VADER sentiment analyzer
analyzer = SentimentIntensityAnalyzer()

# Define a function to classify sentiment based on VADER's compound score.
# Here, we set sentiment 1 (positive) if compound score >= 0.05; otherwise, sentiment 0 (negative).
def get_vader_sentiment(text):
    scores = analyzer.polarity_scores(text)
    return 1 if scores['compound'] >= 0.05 else 0

# Apply the VADER sentiment function to each review in the test set
y_pred_vader = X_test.apply(get_vader_sentiment)

# Evaluate the VADER sentiment predictions
accuracy_vader = accuracy_score(y_test, y_pred_vader)
print("VADER Sentiment Analysis Accuracy:", accuracy_vader)
print("Classification Report:\n", classification_report(y_test, y_pred_vader))


VADER Sentiment Analysis Accuracy: 0.8668783308765167
Classification Report:
               precision    recall  f1-score   support

           0       0.92      0.59      0.72      2554
           1       0.85      0.98      0.91      6265

    accuracy                           0.87      8819
   macro avg       0.89      0.79      0.82      8819
weighted avg       0.87      0.87      0.86      8819



#Observations
Naive Bayes (CountVectorizer)
Accuracy: 92.1%
Precision/Recall Breakdown:
Negative Reviews (Class 0): 87% Precision, 86% Recall
Positive Reviews (Class 1): 94% Precision, 95% Recall
Observation: Naive Bayes performs well, especially in predicting positive sentiments. However, it slightly struggles with misclassifying negative reviews.

Support Vector Machine (SVM) with CountVectorizer
Performance details not extracted, but SVM models usually outperform Naive Bayes in classification tasks where feature sparsity is handled effectively.
Expectation: SVM should have slightly higher accuracy and better balance in recall for negative reviews.

Naive Bayes (TF-IDF)
Expected to perform similarly to the CountVectorizer version, but TF-IDF might give a slight advantage in capturing key words rather than frequency-based importance.
Works well when distinguishing sentiment using unique descriptive words.


SVM (TF-IDF)
Likely the best performer because SVM works well with TF-IDF representations where the focus is on weighted word importance rather than raw counts.
Generally achieves higher accuracy and recall than Naive Bayes.

observations on VADER Performance:
Lexicon-based approaches struggle with context and sarcasm.
The accuracy is expected to be lower than ML models, especially when dealing with complex review texts.
VADER works better in short texts like tweets but not as well for detailed reviews.

Conclusion:

SVM + TF-IDF is expected to be the best performer in this dataset, as SVM works well with high-dimensional text data.
Naive Bayes is fast and efficient but struggles with handling context.
VADER (Lexicon-based) is the least effective as it does not learn from data, making it weaker in real-world sentiment prediction.

Acknowledgement :I referred to ChatGPT for initial guidance and structuring ideas in this assignment; however, the majority of the work, including analysis and implementation, was completed independently without collaboration or reliance on additional tools.

In [41]:
# Install necessary packages
!pip install jupyter nbconvert

# Convert the Jupyter Notebook to an HTML file
!jupyter nbconvert --to html "/content/LA2_maurya_Bhima.ipynb"


[NbConvertApp] Converting notebook /content/LA2_maurya_Bhima.ipynb to html
[NbConvertApp] Writing 307016 bytes to /content/LA2_maurya_Bhima.html
